test/test_torch.py

```diff
@@ -16368,53 +16385,46 @@ def _test(use_out, row_major, incx, incy, lda_tail):
-        dtype = M.dtype
-        numpy_dtype = dtype
-        if dtype in {torch.bfloat16}:
-            numpy_dtype = torch.float
-        if dtype.is_complex:
-            alpha = 0.9 + 0.3j
-            beta = 0.5 + 0.6j
-        else:
-            alpha = 1.2
-            beta = 0.8
-        res1 = torch.addmm(M, m1, m2, alpha=alpha, beta=beta)
-        res2 = torch.full_like(res1, math.nan)
-        torch.addmm(M, m1, m2, alpha=alpha, beta=beta, out=res2)
-        res3 = (beta * M).to(numpy_dtype).cpu().numpy() + alpha * (
-            m1.to(numpy_dtype).cpu().numpy() @ m2.to(numpy_dtype).cpu().numpy())
-        res3 = torch.from_numpy(res3).to(dtype)
-        self.assertEqual(res1, res2)
-        self.assertEqual(res1, res3)
+        for row_major, incx, incy, lda_tail in product((False, True), (1, 2), (1, 2), (0, 1)):
+            _test(row_major, incx, incy, lda_tail)
 
     @precisionOverride({torch.double: 1e-8, torch.float: 1e-4, torch.bfloat16: 0.6,
                         torch.half: 1e-1, torch.cfloat: 1e-4, torch.cdouble: 1e-8})
     @dtypesIfCUDA(*torch.testing.get_all_complex_dtypes(), *torch.testing.get_all_fp_dtypes(include_bfloat16=False))
     @dtypes(*torch.testing.get_all_complex_dtypes(), *torch.testing.get_all_fp_dtypes())
     @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
     def test_addmm(self, device, dtype):
-        M = torch.randn(10, 25).to(device=device, dtype=dtype)
-        m1 = torch.randn(10, 50).to(device=device, dtype=dtype)
-        m2 = torch.randn(50, 25).to(device=device, dtype=dtype)
-        self._test_addmm(M, m1, m2)
+        M = torch.randn(10, 25, device=device).to(dtype)
+        m1 = torch.randn(10, 50, device=device).to(dtype)
+        m2 = torch.randn(50, 25, device=device).to(dtype)
+        self._test_addmm_addmv(torch.addmm, M, m1, m2)
 
         # Test 0-strided
-        M = torch.randn(10, 1).to(device=device, dtype=dtype).expand(10, 25)
-        m1 = torch.randn(10, 1).to(device=device, dtype=dtype).expand(10, 50)
-        m2 = torch.randn(50, 25).to(device=device, dtype=dtype)
-        self._test_addmm(M, m1, m2)
+        M = torch.randn(10, 1, device=device).to(dtype).expand(10, 25)
+        m1 = torch.randn(10, 1, device=device).to(dtype).expand(10, 50)
+        m2 = torch.randn(50, 25, device=device).to(dtype)
+        self._test_addmm_addmv(torch.addmm, M, m1, m2)
+
+        # Test beta=0, M=nan
+        M = torch.full((10, 25), math.nan, device=device).to(dtype)
+        m1 = torch.randn(10, 50, device=device).to(dtype)
+        m2 = torch.randn(50, 25, device=device).to(dtype)
+        self._test_addmm_addmv(torch.addmm, M, m1, m2, beta=0)
+
+        # Test transpose
+        for t1, t2, t3, t4 in product([True, False], repeat=4):
+            M = torch.randn(10, 25, device=device).to(dtype)
+            m1 = torch.randn(10, 50, device=device).to(dtype)
+            m2 = torch.randn(50, 25, device=device).to(dtype)
+            if t1:
```

Comment:
nit: maybe factor it out into `maybe_transpose` function?

Reply:
done